### PR TITLE
perf(semantic): reduce overhead of cfg recording ast nodes

### PR DESCRIPTION
**Pull request by overlookmotel**
Control flow graph builder records AST node IDs in a temp structure `ast_nodes_records`. Only the first AST node ID inserted into the `Vec` is ever used, so turn it into just a single value.

Using `AstNodeId::dummy()` as sentinel for "not set" instead of using `Option<AstNodeId>` to reduce size of the records stack. `Option<AstNodeId>` is 16 bytes.

### CODE REVIEW

**Comment by overlookmotel** [comment]
* **#4263** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4263?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4262** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4262?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by rzvxa — APPROVED** [review]
Nice, I wanted to make this change for a long time, I'm glad you went ahead and refactored this.😍

File: crates/oxc_semantic/src/builder.rs [hunk 1]
```diff
@@ -250,16 +250,19 @@ impl<'a> SemanticBuilder<'a> {
     }
 
     fn record_ast_nodes(&mut self) {
```
Comment (rzvxa) [inline crates/oxc_semantic/src/builder.rs:252]: this should be `record_ast_node` now. Maybe we can rename the other method to something like `push_ast_record`?

Comment (rzvxa) [reply crates/oxc_semantic/src/builder.rs:252]: @Boshen Feel free to resolve my nitpicks. None of my comments are a blocking change, Just suggestions.

**Pull request merged**